sycl/unittests/SYCL2020/IsCompatible.cpp

```diff
@@ -0,0 +1,234 @@
+
+  return Img;
+}
+
+static sycl::unittest::PiImage Imgs[3] = {
+    generateDefaultImage({"TestKernelCPU"}, {sycl::aspect::cpu}),
+    generateDefaultImage({"TestKernelGPU"}, {sycl::aspect::gpu}),
+    generateDefaultImage({"TestKernelACC"}, {sycl::aspect::accelerator})};
+
+static sycl::unittest::PiImageArray<3> ImgArray{Imgs};
+
+static pi_result redefinedDeviceGetInfoCPU(pi_device device,
+                                           pi_device_info param_name,
+                                           size_t param_value_size,
+                                           void *param_value,
+                                           size_t *param_value_size_ret) {
+  if (param_name == PI_DEVICE_INFO_TYPE) {
+    auto *Result = reinterpret_cast<_pi_device_type *>(param_value);
+    *Result = PI_DEVICE_TYPE_CPU;
+  }
+  return PI_SUCCESS;
+}
+
+// Mock device is "GPU" by default, but we need to redefine it just in case
+// if there are some changes in the future
+static pi_result redefinedDeviceGetInfoGPU(pi_device device,
+                                           pi_device_info param_name,
+                                           size_t param_value_size,
+                                           void *param_value,
+                                           size_t *param_value_size_ret) {
+  if (param_name == PI_DEVICE_INFO_TYPE) {
+    auto *Result = reinterpret_cast<_pi_device_type *>(param_value);
+    *Result = PI_DEVICE_TYPE_GPU;
+  }
+  return PI_SUCCESS;
+}
+
+static pi_result redefinedDeviceGetInfoACC(pi_device device,
+                                           pi_device_info param_name,
+                                           size_t param_value_size,
+                                           void *param_value,
+                                           size_t *param_value_size_ret) {
+  if (param_name == PI_DEVICE_INFO_TYPE) {
+    auto *Result = reinterpret_cast<_pi_device_type *>(param_value);
+    *Result = PI_DEVICE_TYPE_ACC;
+  }
+  return PI_SUCCESS;
+}
+
+TEST(IsCompatible, CPU) {
+  sycl::unittest::PiMock Mock;
+  Mock.redefineAfter<sycl::detail::PiApiKind::piDeviceGetInfo>(
+      redefinedDeviceGetInfoCPU);
+  sycl::platform Plt = Mock.getPlatform();
+  const sycl::device Dev = Plt.get_devices()[0];
+  sycl::queue Q(Dev);
+  bool Compatible = true;
+  bool Called = false;
+
+  if (sycl::is_compatible<TestKernelCPU>(Dev)) {
+    Q.submit([&](sycl::handler &h) { h.single_task<TestKernelCPU>([=]() {}); });
+    Q.wait();
+    Compatible &= Dev.is_cpu();
+    Called = true;
+  }
+  if (sycl::is_compatible<TestKernelGPU>(Dev)) {
+    Q.submit([&](sycl::handler &h) { h.single_task<TestKernelGPU>([=]() {}); });
+    Q.wait();
+    Compatible &= Dev.is_gpu();
+    Called = true;
+  }
+  if (sycl::is_compatible<TestKernelACC>(Dev)) {
+    Q.submit([&](sycl::handler &h) { h.single_task<TestKernelACC>([=]() {}); });
+    Q.wait();
+    Compatible &= Dev.is_accelerator();
+    Called = true;
+  }
```

Comment:
Since we have full control over the devices, shouldn't the non-CPU branches be unexpected and fail?

Reply:
Do you mean if "is_compatible" returns true, but device really doesn't support this kernel feature?

Reply:
By the way, if I understand the unit-tests essence correctly, we don't really need to define queue since the kernel is already defined. Ok, I'll fix it

Reply:
Does it look better now? [3ddc3d1](https://github.com/intel/llvm/pull/7619/commits/3ddc3d15418e1f16e9abf92f5a2f9efd5d0f03ae)

Reply:
I think so! Thanks!